src/charm.py

```diff
@@ -1,104 +1,124 @@
+from ops.charm import CharmBase, WorkloadEvent
 from ops.main import main
-from ops.model import ActiveStatus
+from ops.model import ActiveStatus, WaitingStatus
+from ops.pebble import Layer
 
 logger = logging.getLogger(__name__)
 
 
-class OperatorTemplateCharm(CharmBase):
-    """Charm the service."""
-
-    _stored = StoredState()
+class MysqlOperatorCharm(CharmBase):
+    """A Juju Charm for MySQL InnoDB Cluster."""
 
     def __init__(self, *args):
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
-        self.framework.observe(self.on.config_changed, self._on_config_changed)
-        self.framework.observe(self.on.fortune_action, self._on_fortune_action)
-        self._stored.set_default(things=[])
 
-    def _on_httpbin_pebble_ready(self, event):
-        """Define and start a workload using the Pebble API.
+        self._config = self.model.config
 
-        TEMPLATE-TODO: change this example to suit your needs.
-        You'll need to specify the right entrypoint and environment
-        configuration for your specific workload. Tip: you can see the
-        standard entrypoint of an existing container using docker inspect
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        change_events = [
+            self.on.mysql_server_pebble_ready,
+            self.on.mysql_router_pebble_ready,
+        ]
+        for event in change_events:
+            self.framework.observe(event, self._on_pebble_ready)
+
+    # ---- Event handlers
+    def _on_start(self, _):
+        """Placeholder for StartEvent."""
+        pass
 
-        Learn more about Pebble layers at https://github.com/canonical/pebble
-        """
-        # Get a reference the container attribute on the PebbleReadyEvent
+    def _on_config_changed(self, _):
+        """Placeholder for ConfigChangedEvent."""
+        pass
+
+    def _on_pebble_ready(self, event: WorkloadEvent) -> None:
+        """Updates the Pebble configuration layer when changed."""
         container = event.workload
-        # Define an initial Pebble layer configuration
-        pebble_layer = {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
+        if not container.can_connect():
+            self.unit.status = WaitingStatus("Waiting for pod startup to complete")
+            event.defer()
+            return
+
+        # Get current config
+        current_layer = container.get_plan()
+        # Create a new config layer
+        service = container.name
+        new_layer = self._mysql_layer(service=service)
+
+        if current_layer.services != new_layer.services:
+            container.add_layer(service, new_layer, combine=True)
+            logging.info("Pebble plan updated with new configuration")
+            container.restart(service)
+
```

Comment:
Yeh, I agree that something could go wrong with the service restart. It could be something like:

```python
try:
    container.restart(service_name)
except ops.pebble.ChangeError as e:
    <something, maybe BlockedStatus?>
else:
    self.unit.status = ActiveStatus()
```

I haven't seen examples of error handling for the restart, so I'm not sure. @jnsgruk thoughts?

Reply:
The answer to that question depends on whether the failure is something that you can recover automatically from.

If you can recover, execute the recovery code (which might just look like deferring the event, possibly tracking how many times that you have done so in self._stored, so that you can give up after some time).

If you can't recover, and a human needs to intervene, set BlockedStatus and return.

Note that if you don't catch the error, the hook will fail, and the charm will go into error status, re-running the hook until the problem fixes itself, or a human intervenes.

Reply:
Note: implicit in the answer above is that you have some understanding of the underlying workload. What can go wrong when you're setting up MySQL? What sort of errors will you get back? Is the service ready immediately after the restart, or do you need to do a liveness check? 

This is the heart of what charms are: a framework that makes it easy to focus just on the bits unique to the workload, with code that handles common failure modes, and best practices around handling those failures.